project_cost_report/report.py

```diff
@@ -0,0 +1,436 @@
+# © 2019 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from odoo import api, models, _
+from odoo.tools.float_utils import float_round
+from odoo.exceptions import ValidationError
+from typing import Callable, Mapping
+
+
+class ProjectCostReport(models.TransientModel):
+    """Project Cost Report.
+
+    This model is used to generate the cost report.
+
+    The report is structured as follow:
+
+    SECTION 1
+    ---------
+
+    Category 1.1
+    ~~~~~~~~~~~~
+    Analytic Line 1.1.1
+    Analytic Line 1.1.2
+
+    Category 1.2
+    ~~~~~~~~~~~~
+    Analytic Line 1.2.1
+
+    SECTION 2
+    ---------
+
+    Category 2.1
+    ~~~~~~~~~~~~
+    Analytic Line 2.1.1
+    ...
+
+    The report sections are static (defined in the code):
+
+    * PRODUCTS
+    * TIME
+    * OUTSOURCING
+    * etc
+
+    The categories are dynamic.
+    The grouping key of each category is different:
+
+    * Categories under PRODUCTS are product categories.
+    * Categories under TIME are task types.
+    * OUTSOURCING has only one category (Outsourcing).
+
+    Each category contains a given recordset of analytic lines.
+
+    If a report category has no analytic line, it is not shown.
+    """
+
+    _name = 'project.cost.report'
+    _description = 'Project Cost Report'
+
+    def _get_rendering_variables(self, project, report_context):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=3675949)